src/bucket/BucketManagerImpl.cpp

```diff
@@ -867,4 +868,165 @@ BucketManagerImpl::isShutdown() const
+    double mb = static_cast<double>(sz) / 1000000.0;
+    double ms = timer.max();
+    double sec = ms / 1000.0;
+    CLOG_INFO(Bucket, "{} {}-byte bucket file in {:.2f}ms ({:.2f} MB/s)", pass,
+              sz, ms, mb / sec);
+}
+
+bool
+BucketManagerImpl::selfCheck(
+    HistoryArchiveState const& has,
+    UnorderedMap<LedgerKey, LedgerEntry>& ledgerMap) const
+{
+    bool ok = true;
+
+    std::vector<std::string> files =
+        fs::findfiles(getBucketDir(), isBucketFile);
+    std::set<Hash> hashesInDir;
+    for (auto const& f : files)
+    {
+        hashesInDir.insert(extractFromFilename(f));
+    }
+
+    // First we collect the set and order of buckets we are expecting to see.
+    std::set<Hash> hashesInBl;
+    std::set<Hash> hashesInFuture;
+    std::vector<Hash> hashOrderToVisit;
+    for (uint32_t i = BucketList::kNumLevels; i > 0; --i)
+    {
+        HistoryStateBucket const& hsb = has.currentBuckets.at(i - 1);
+        Hash snap = hexToBin256(hsb.snap);
+        Hash curr = hexToBin256(hsb.curr);
+        hashesInBl.insert(snap);
+        hashesInBl.insert(curr);
+        std::vector<Hash> expected = {snap, curr};
+        if (hsb.next.hasOutputHash())
+        {
+            Hash next = hexToBin256(hsb.next.getOutputHash());
+            hashesInFuture.insert(next);
+            expected.emplace_back(next);
+        }
+        for (auto const& e : expected)
+        {
+            if (isZero(e))
+            {
+                continue;
+            }
+            hashOrderToVisit.emplace_back(e);
+            if (hashesInDir.find(e) == hashesInDir.end())
+            {
+                CLOG_ERROR(Bucket,
+                           "Missing expected level {} file for bucket {}",
+                           i - 1, binToHex(e));
+                ok = false;
+            }
+            else
+            {
+                CLOG_INFO(Bucket, "Found expected level {} file for bucket {}",
+                          i - 1, binToHex(e));
+            }
+        }
+    }
+
+    for (auto const& h : hashesInDir)
+    {
+        if (hashesInBl.find(h) == hashesInBl.end() &&
+            hashesInFuture.find(h) == hashesInFuture.end())
+        {
+            // This is a hash not-in the BL or a FutureBucket. Warn, but not
+            // error.
+            CLOG_WARNING(Bucket, "found unexpected bucket {}", binToHex(h));
+            hashOrderToVisit.emplace_back(h);
+        }
+    }
+
+    // Then we scan the existing buckets, collecting entries from those that are
+    // relevant to the current ledger state into the provided ledger map.
+    //
+    // We do this in 3 passes so we can get relatively-useful isolated IO and
+    // hashing performance metrics out of the first 2 passes (to get a hint if
+    // the user's machine is slow in some important way).
+    for (auto h : hashOrderToVisit)
+    {
+        auto path = bucketFilename(h);
+        size_t sz = fs::size(path);
+
+        auto b = std::make_shared<Bucket>(path, h);
```

Comment:
Mainly because I was hoping to keep `selfCheck` const. Whether that's a good goal or not, I didn't see it making much difference -- it's a 1-line call either way.

Reply:
(Decided to go with your suggestion -- const isn't really important here and it reads a little lighter)